python/lsst/daf/persistence/butler.py

```diff
@@ -136,44 +175,213 @@ def __init__(self, root, mapper=None, **mapperArgs):
+                    # repositories used by this butler. (If there is a configuration change a new output 
+                    # repository should be created). 
+                    # IE cfg.parents can have parents that are not passed in as butler inputs or readable 
+                    # outputs, but the butler may not have readable i/o that is not a parent of an already 
+                    # existing output.
+                    for cfgRoot in butlerIOParents:
+                        if cfgRoot not in cfg.parents:
+                            if cfgRoot != args.cfgRoot or 'r' not in args.mode:
+                                raise RuntimeError(
+                                    "Existing output repository parents do not match butler's inputs.")
+                if not cfg.matchesArgs(args):
+                    raise RuntimeError(
+                       "Persisted RepositoryCfg and passed-in RepositoryArgs have conflicting parameters:\n" +
+                       "\t%s\n\t%s", (cfg, args))
+                if args.mapperArgs is not None:
+                    if cfg.mapperArgs is None:
+                        cfg.mapperArgs = args.mapperArgs
+                    else:
+                        cfg.mapperArgs.update(args.mapperArgs)
+                if 'r' in args.mode:
+                    parentsToAdd = copy.copy(cfg.parents)
+            else:
+                if args.mapper is None:
+                    if defaultMapper is None:
+                        raise RuntimeError(
+                            "Could not infer mapper and one not specified in repositoryArgs:%s" % args)
+                    args.mapper = defaultMapper
+                parents = [cfgRoot for cfgRoot in butlerIOParents.keys() if cfgRoot != args.cfgRoot]
+                cfg = RepositoryCfg.makeFromArgs(args, parents)
+                Storage.putRepositoryCfg(cfg, args.cfgRoot)
+
+            repo = Repository(cfg)
+            self._repos.add(RepoData(args, cfg, repo, tags))
+            for parent in parentsToAdd:
+                if parent in butlerIOParents:
+                    args = butlerIOParents[parent]
+                else:
+                    args = RepositoryArgs(cfgRoot=parent, mode='r')
+                self._addRepo(args=args, inout='in', tags=tags)
 
 
     def __repr__(self):
-        return 'Butler(cfg=%s, datasetTypeAliasDict=%s, repository=%s, persistence=%s)' % (
-            self._cfg, self.datasetTypeAliasDict, self.repository, self.persistence)
+        return 'Butler(datasetTypeAliasDict=%s, repos=%s, persistence=%s)' % (
+            self.datasetTypeAliasDict, self._repos, self.persistence)
 
+    @staticmethod
+    def getDefaultMapper(inputs):
+        mappers = set()
+        for args in inputs:
+            if args.mapper is not None:
+                mappers.add(args.mapper)
+            else:
+                cfgRoot = args.cfgRoot
+                mapper = Butler.getMapperClass(cfgRoot)
+                mappers.add(mapper)
```

Comment:
If I'm reading this right, `args.mapper` is either unsanitized or an instance at this stage, which I don't think we can expect to be hashable, right?  Or is that a new requirement on mappers?